File: scripts/backfill_notion_impact.py
import json
import requests
import datetime
from dateutil import parser
import os

NOTION_KEY = os.popen('cat ~/.config/notion/api_key').read().strip()
NOTION_VERSION = "2025-09-03"
DATA_SOURCE_ID = "31685ec7-0668-815a-bc98-000bab1964f3"
NS_URL = "https://p01--sefi--s66fclg7g2lm.code.run"

headers = {
    "Authorization": f"Bearer {NOTION_KEY}",
    "Notion-Version": NOTION_VERSION,
    "Content-Type": "application/json"
}

def get_ns_entries(count=1000):
    r = requests.get(f"{NS_URL}/api/v1/entries.json?count={count}")
    return r.json()

def get_notion_items():
    results = []
    has_more = True
    next_cursor = None
    while has_more:
        payload = {"page_size": 100}
        if next_cursor:
            payload["start_cursor"] = next_cursor
        r = requests.post(f"https://api.notion.com/v1/data_sources/{DATA_SOURCE_ID}/query", headers=headers, json=payload)
        data = r.json()
        results.extend(data.get("results", []))
        has_more = data.get("has_more", False)
        next_cursor = data.get("next_cursor", None)
    return results

def get_bg_at(entries, timestamp):
    target = timestamp.timestamp() * 1000
    closest = None
    min_diff = 10 * 60 * 1000 # 10 minutes
    for e in entries:
        diff = abs(e['mills'] - target)
        if diff < min_diff:
            min_diff = diff
            closest = e
    return closest['sgv'] if closest else None

def get_peak_2hr(entries, meal_ts):
    start = meal_ts.timestamp() * 1000
    end = (meal_ts + datetime.timedelta(hours=3.5)).timestamp() * 1000 # Extended search window
    peak_bg = 0
    peak_time_ms = None
    for e in entries:
        if start <= e['mills'] <= end:
            if e['sgv'] > peak_bg:
                peak_bg = e['sgv']
                peak_time_ms = e['mills']
    
    peak_ts = datetime.datetime.fromtimestamp(peak_time_ms / 1000, tz=datetime.timezone.utc) if peak_time_ms else None
    return peak_bg, peak_ts

ns_entries = get_ns_entries()
notion_items = get_notion_items()

print(f"Checking {len(notion_items)} Notion items...")

for item in notion_items:
    props = item['properties']
    category = props.get('Category', {}).get('select', {}).get('name')
    if category != "Food":
        continue
    
    bg_delta_prop = props.get('BG Delta', {})
    bg_delta_val = bg_delta_prop.get('number')
    
    # Check if ANY of the impact columns are missing
    pre_bg_prop = props.get('Pre-Meal BG', {}).get('number')
    peak_bg_prop = props.get('2hr Peak BG', {}).get('number')
    
    if bg_delta_val is not None and pre_bg_prop is not None and peak_bg_prop is not None:
        continue # Already fully processed
        
    date_str = props.get('Date', {}).get('date', {}).get('start')
    if not date_str:
        continue
    
    meal_ts = parser.parse(date_str)
    # Correct for local time vs UTC in Nightscout if needed
    # If the date_str has no timezone, assume it's UTC or the local time based on the script
    
    pre_bg = get_bg_at(ns_entries, meal_ts)
    peak_bg, peak_ts = get_peak_2hr(ns_entries, meal_ts)
    
    if pre_bg and peak_bg:
        delta = peak_bg - pre_bg
        time_to_peak = int((peak_ts - meal_ts).total_seconds() / 60) if peak_ts else None
        
        entry_title = props.get('Entry', {}).get('title', [])
        title_text = entry_title[0]['text']['content'] if entry_title else "Untitled"
        
        update_payload = {
            "properties": {
                "Pre-Meal BG": {"number": pre_bg},
                "2hr Peak BG": {"number": peak_bg},
                "BG Delta": {"number": delta},
            }
        }
        if peak_ts:
            update_payload["properties"]["Peak Time"] = {"date": {"start": peak_ts.isoformat()}}
        if time_to_peak is not None:
            update_payload["properties"]["Time to Peak (min)"] = {"number": time_to_peak}
            
        print(f"Updating '{title_text}' ({meal_ts}): Pre {pre_bg}, Peak {peak_bg}, Delta {delta}")
        r = requests.patch(f"https://api.notion.com/v1/pages/{item['id']}", headers=headers, json=update_payload)
        if r.status_code != 200:
            print(f"Error updating {item['id']}: {r.text}")

File: scripts/backfill_notion_impact.js
const https = require('https');
const fs = require('fs');

const NOTION_KEY = "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";
const DATA_SOURCE_ID = "31685ec7-0668-813e-8b9e-c5b4d5d70fa5";
const NS_URL = "https://p01--sefi--s66fclg7g2lm.code.run";
const NS_SECRET_HASH = "b3170e23f45df7738434cd8be9cd79d86a6d0f01"; // SHA1 of JaviCare2026

async function fetchJson(url, headers = {}) {
  return new Promise((resolve, reject) => {
    const options = { headers };
    if (url.includes('code.run') && !headers['api-secret']) {
      options.headers = { ...options.headers, 'api-secret': NS_SECRET_HASH };
    }
    https.get(url, options, (res) => {
      let data = ''; 
      res.on('data', (c) => data += c); 
      res.on('end', () => {
        try {
          resolve(JSON.parse(data || '[]'));
        } catch (e) {
          console.error(`Error parsing JSON from ${url}:`, e.message);
          console.error('Response starts with:', data.substring(0, 100));
          reject(e);
        }
      });
    }).on('error', reject);
  });
}

async function postJson(url, payload) {
  return new Promise((resolve, reject) => {
    const data = JSON.stringify(payload);
    const options = {
      method: 'POST', headers: { 'Authorization': 'Bearer ' + NOTION_KEY, 'Notion-Version': '2022-06-28', 'Content-Type': 'application/json', 'Content-Length': Buffer.byteLength(data) }
    };
    const req = https.request(url, options, (res) => { 
      let body = ''; 
      res.on('data', (c) => body += c); 
      res.on('end', () => {
        try {
          resolve(JSON.parse(body || '{}'));
        } catch (e) {
          console.error(`Error parsing JSON from POST ${url}:`, e.message);
          console.error('Response starts with:', body.substring(0, 100));
          reject(e);
        }
      });
    });
    req.on('error', reject);
    req.write(data); req.end();
  });
}

async function patchJson(url, payload) {
  return new Promise((resolve, reject) => {
    const data = JSON.stringify(payload);
    const options = {
      method: 'PATCH', headers: { 'Authorization': 'Bearer ' + NOTION_KEY, 'Notion-Version': '2022-06-28', 'Content-Type': 'application/json', 'Content-Length': Buffer.byteLength(data) }
    };
    const req = https.request(url, options, (res) => { 
      let body = ''; 
      res.on('data', (c) => body += c); 
      res.on('end', () => {
        try {
          resolve(JSON.parse(body || '{}'));
        } catch (e) {
          console.error(`Error parsing JSON from PATCH ${url}:`, e.message);
          console.error('Response starts with:', body.substring(0, 100));
          reject(e);
        }
      });
    });
    req.on('error', reject);
    req.write(data); req.end();
  });
}

function getBgAt(entries, mealTime) {
  const target = new Date(mealTime).getTime();
  let closest = null;
  let minDiff = 30 * 60 * 1000;
  for (const e of entries) {
    const mills = e.date || e.mills;
    const diff = Math.abs(mills - target);
    if (diff < minDiff) {
      minDiff = diff;
      closest = e;
    }
  }
  return closest ? closest.sgv : null;
}

function getPeak2Hr(entries, mealTime) {
  const start = new Date(mealTime).getTime();
  const end = start + 3 * 60 * 60 * 1000;
  let peakBg = 0;
  let peakTimeMs = null;
  for (const e of entries) {
    const mills = e.date || e.mills;
    if (mills >= start && mills <= end) {
      if (e.sgv > peakBg) {
        peakBg = e.sgv;
        peakTimeMs = mills;
      }
    }
  }
  return { peakBg, peakTime: peakTimeMs ? new Date(peakTimeMs).toISOString() : null };
}

async function run() {
  console.log('Starting Impact Variance Audit...');
  // 576 = 48h of CGM at 5-min intervals; enough for any backfill window
  const nsEntries = await fetchJson(`${NS_URL}/api/v1/entries.json?count=576`);
  
  let results = [];
  let hasMore = true;
  let cursor = undefined;

  while (hasMore) {
    const res = await postJson(`https://api.notion.com/v1/databases/${DATA_SOURCE_ID}/query`, { 
      filter: { property: 'Date', date: { on_or_after: '2026-03-06' } },
      start_cursor: cursor
    });
    
    if (res.results) {
      results = results.concat(res.results);
    }
    hasMore = res.has_more;
    cursor = res.next_cursor;
    console.log(`Fetched ${results.length} entries from Notion...`);
  }

  for (const item of results) {
    if (item.archived) continue;
    const props = item.properties;
    if (props.Category.select.name !== 'Food') continue;
    
    const titleText = props.Entry.title[0]?.plain_text;
    const dateStr = props.Date.date.start;
    
    // MATURITY CHECK: Only process meals that are at least 3 hours old
    const mealTime = new Date(dateStr);
    const now = new Date();
    const ageInHours = (now - mealTime) / (1000 * 60 * 60);
    
    if (ageInHours < 3) {
      console.log(`Skipping '${titleText}' (${dateStr}): Too recent (age: ${ageInHours.toFixed(1)}h)`);
      continue;
    }

    const updatePayload = { properties: {} };

    // 1. Ensure real peak data is current
    let currentPeakBg = props['2hr Peak BG']?.number;
    let currentPeakTimeStr = props['Peak Time']?.date?.start;
    let currentPreBg = props['Pre-Meal BG']?.number;

    if (!currentPeakBg || !currentPreBg) {
        const preBg = getBgAt(nsEntries, dateStr);
        const { peakBg, peakTime } = getPeak2Hr(nsEntries, dateStr);
        if (preBg && peakBg) {
            currentPreBg = preBg;
            currentPeakBg = peakBg;
            currentPeakTimeStr = peakTime;
            const delta = peakBg - preBg;
            const timeToPeak = Math.round((new Date(peakTime) - new Date(dateStr)) / (1000 * 60));
            
            updatePayload.properties['Pre-Meal BG'] = { number: preBg };
            updatePayload.properties['2hr Peak BG'] = { number: peakBg };
            updatePayload.properties['BG Delta'] = { number: delta };
            updatePayload.properties['Peak Time'] = { date: { start: peakTime } };
            updatePayload.properties['Time to Peak (min)'] = { number: timeToPeak };
        }
    }

    // 2. Calculate Variances if predictions exist
    const predPeakBg = props['Predicted Peak BG']?.number;
    const predPeakTimeStr = props['Predicted Peak Time']?.date?.start;

    if (predPeakBg != null && currentPeakBg != null) {
      const bgVar = currentPeakBg - predPeakBg;
      updatePayload.properties['Peak BG Delta'] = { number: bgVar };
      
      if (currentPeakTimeStr && predPeakTimeStr) {
        const predDate = new Date(predPeakTimeStr);
        const peakDate = new Date(currentPeakTimeStr);
        const timeVar = Math.round((peakDate - predDate) / (1000 * 60));
        updatePayload.properties['Peak Time Delta (min)'] = { number: timeVar };
      }
    }

    if (Object.keys(updatePayload.properties).length > 0) {
        console.log(`Updating '${titleText}' (${dateStr}): Pre ${currentPreBg}, Peak ${currentPeakBg}`);
        await patchJson(`https://api.notion.com/v1/pages/${item.id}`, updatePayload);
    }
  }
  console.log('Variance Audit Complete.');
}
run();
